Compilation Fails

Compiling phylanx on a Raspberry Pi/Fedora fails with the following error:
```
[ 16%] Building CXX object src/CMakeFiles/phylanx_component.dir/execution_tree/primitives/primitive_component_base.cpp.o
[ 16%] Building CXX object src/CMakeFiles/phylanx_component.dir/execution_tree/primitives/slice.cpp.o
[ 18%] Building CXX object src/CMakeFiles/phylanx_component.dir/execution_tree/primitives/slice_node_data.cpp.o
In file included from /home/a/phylanx/src/execution_tree/primitives/slice_node_data.cpp:12:
/home/a/phylanx/phylanx/execution_tree/primitives/slice_node_data_2d.hpp:269:13: error: 
      static_assert failed "sizeof(std::int64_t) == sizeof(std::size_t)"
            static_assert(sizeof(std::int64_t) == sizeof(std::size_t),
            ^             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/a/phylanx/phylanx/execution_tree/primitives/slice_node_data_2d.hpp:453:13: error: 
      static_assert failed "sizeof(std::int64_t) == sizeof(std::size_t)"
            static_assert(sizeof(std::int64_t) == sizeof(std::size_t),
            ^             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
2 errors generated.
make[2]: *** [src/CMakeFiles/phylanx_component.dir/build.make:544: src/CMakeFiles/phylanx_component.dir/execution_tree/primitives/slice_node_data.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:1133: src/CMakeFiles/phylanx_component.dir/all] Error 2
make: *** [Makefile:141: all] Error 2
```
Configuring the compilation with: -DHPX_DIR=/opt/hpx -DCMAKE_INSTALL_PREFIX=/opt/phylanx -DHPX_IGNORE_COMPILER_COMPATIBILITY=On
and calling make with -j 1.

The output of uname -a is:
```
Linux localhost 4.20.10-200.fc29.armv7hl #1 SMP Fri Feb 15 21:04:49 UTC 2019 armv7l armv7l armv7l GNU/Linux
```
CMakeFiles/CMakeError.log contains the following information:
```
Performing C++ SOURCE FILE Test HAS_FLTO failed with the following output:
Change Dir: /home/a/phylanx/build/CMakeFiles/CMakeTmp

Using clang++ version:
clang version 7.0.1 (Fedora 7.0.1-2.fc29)

Run Build Command:"/usr/bin/gmake" "cmTC_7e8a7/fast"
/usr/bin/gmake -f CMakeFiles/cmTC_7e8a7.dir/build.make CMakeFiles/cmTC_7e8a7.dir/build
gmake[1]: Entering directory '/home/a/phylanx/build/CMakeFiles/CMakeTmp'
Building CXX object CMakeFiles/cmTC_7e8a7.dir/src.cxx.o
/usr/bin/clang++    -fdiagnostics-show-option -Werror=vla -Werror=return-type -Wno-cast-align -DHAS_FLTO   -flto -o CMakeFiles/cmTC_7e8a7.dir/src.cxx.o -c /home/a/phylanx/build/CMakeFiles/CMakeTmp/src.cxx
Linking CXX executable cmTC_7e8a7
/usr/bin/cmake -E cmake_link_script CMakeFiles/cmTC_7e8a7.dir/link.txt --verbose=1
/usr/bin/clang++   -fdiagnostics-show-option -Werror=vla -Werror=return-type -Wno-cast-align -DHAS_FLTO   -Wl,-z,defs  -rdynamic CMakeFiles/cmTC_7e8a7.dir/src.cxx.o  -o cmTC_7e8a7 -flto 
/usr/bin/ld: BFD version 2.31.1-17.fc29 internal error, aborting at elfcode.h:224 in bfd_elf32_swap_symbol_out

/usr/bin/ld: Please report this bug.

clang-7: error: linker command failed with exit code 1 (use -v to see invocation)
gmake[1]: *** [CMakeFiles/cmTC_7e8a7.dir/build.make:87: cmTC_7e8a7] Error 1
gmake[1]: Leaving directory '/home/a/phylanx/build/CMakeFiles/CMakeTmp'
gmake: *** [Makefile:121: cmTC_7e8a7/fast] Error 2
```
Source file was:
int main() { return 0; }

After applying the 814.patch the compilation fails at a different place with this output
/opt/hpx/include/hpx/util/spinlock_pool.hpp:87:59: note: in instantiation
      of member function
      'hpx::util::spinlock_pool<hpx::naming::gid_type::tag,
      128>::spinlock_for' requested here
  ...explicit scoped_lock( void const * pv ): sp_( spinlock_for( pv ) )
                                                   ^
/opt/hpx/include/hpx/runtime/naming/name.hpp:383:46: note: in
      instantiation of member function
      'hpx::util::spinlock_pool<hpx::naming::gid_type::tag,
      128>::scoped_lock::scoped_lock' requested here
            internal_mutex_type::scoped_lock l(this);
                                             ^
In file included from /home/a/phylanx/src/plugins/matrixops/softmax_operation.cpp:8:
In file included from /home/a/phylanx/phylanx/ir/node_data.hpp:13:
In file included from /opt/hpx/include/hpx/include/serialization.hpp:10:
In file included from /opt/hpx/include/hpx/runtime/serialization/serialize.hpp:12:
In file included from /opt/hpx/include/hpx/runtime/serialization/input_archive.hpp:16:
In file included from /opt/hpx/include/hpx/runtime/serialization/input_container.hpp:12:
In file included from /opt/hpx/include/hpx/runtime/serialization/container.hpp:12:
In file included from /opt/hpx/include/hpx/runtime/naming/name.hpp:12:
In file included from /opt/hpx/include/hpx/util/spinlock_pool.hpp:23:
/opt/hpx/include/hpx/util/fibhash.hpp:50:13: warning: shift count >= width
      of type [-Wshift-count-overflow]
            >> helper::shift_amount;
            ^  ~~~~~~~~~~~~~~~~~~~~
3 warnings and 4 errors generated.
make[2]: *** [src/plugins/matrixops/CMakeFiles/matrixops_primitive.dir/build.make:492: src/plugins/matrixops/CMakeFiles/matrixops_primitive.dir/softmax_operation.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:2056: src/plugins/matrixops/CMakeFiles/matrixops_primitive.dir/all] Error 2
make: *** [Makefile:141: all] Error 2

and the relevant part in the CMakeError.log:

Performing C++ SOURCE FILE Test HAS_FLTO failed with the following output:
Change Dir: /home/a/phylanx/build/CMakeFiles/CMakeTmp

Run Build Command:"/usr/bin/gmake" "cmTC_2fb84/fast"
/usr/bin/gmake -f CMakeFiles/cmTC_2fb84.dir/build.make CMakeFiles/cmTC_2fb84.dir/build
gmake[1]: Entering directory '/home/a/phylanx/build/CMakeFiles/CMakeTmp'
Building CXX object CMakeFiles/cmTC_2fb84.dir/src.cxx.o
/usr/bin/clang++    -fdiagnostics-show-option -Werror=vla -Werror=return-type -Wno-cast-align -DHAS_FLTO   -flto -o CMakeFiles/cmTC_2fb84.dir/src.cxx.o -c /home/a/phylanx/build/CMakeFiles/CMakeTmp/src.cxx
Linking CXX executable cmTC_2fb84
/usr/bin/cmake -E cmake_link_script CMakeFiles/cmTC_2fb84.dir/link.txt --verbose=1
/usr/bin/clang++   -fdiagnostics-show-option -Werror=vla -Werror=return-type -Wno-cast-align -DHAS_FLTO   -Wl,-z,defs  -rdynamic CMakeFiles/cmTC_2fb84.dir/src.cxx.o  -o cmTC_2fb84 -flto 
/usr/bin/ld: BFD version 2.31.1-17.fc29 internal error, aborting at elfcode.h:224 in bfd_elf32_swap_symbol_out

/usr/bin/ld: Please report this bug.

clang-7: error: linker command failed with exit code 1 (use -v to see invocation)
gmake[1]: *** [CMakeFiles/cmTC_2fb84.dir/build.make:87: cmTC_2fb84] Error 1
gmake[1]: Leaving directory '/home/a/phylanx/build/CMakeFiles/CMakeTmp'
gmake: *** [Makefile:121: cmTC_2fb84/fast] Error 2

Source file was:
int main() { return 0; }

--- a/phylanx/execution_tree/primitives/slice_node_data_2d.hpp
+++ b/phylanx/execution_tree/primitives/slice_node_data_2d.hpp
@@ -18,9 +18,12 @@
 #include <hpx/throw_exception.hpp>
 #include <hpx/util/assert.hpp>
 
+#include <algorithm>
 #include <cstddef>
 #include <cstdint>
+#include <memory>
 #include <string>
+#include <type_traits>
 #include <utility>
 
 #include <blaze/Math.h>
@@ -183,6 +186,48 @@ namespace phylanx { namespace execution_tree
     }
 
     ///////////////////////////////////////////////////////////////////////////
+    namespace detail
+    {
+        template <typename T, typename Index, typename Enable = void>
+        struct extract_columns;
+
+        template <typename T, typename Index>
+        struct extract_columns<T, Index,
+            typename std::enable_if<sizeof(std::int64_t) == sizeof(Index)>::type>
+        {
+            template <typename F, typename Data, typename SubMatrix>
+            ir::node_data<T> operator()(F const& f, Data&& m, SubMatrix&& sm,
+                ir::node_data<std::int64_t>&& columns) const
+            {
+                auto result = blaze::columns(sm,
+                    reinterpret_cast<std::size_t*>(columns.vector().data()),
+                    columns.size());
+
+                return f.matrix(m, std::move(result));
+            }
+        };
+
+        template <typename T, typename Index>
+        struct extract_columns<T, Index,
+            typename std::enable_if<sizeof(std::int64_t) != sizeof(Index)>::type>
+        {
+            template <typename F, typename Data, typename SubMatrix>
+            ir::node_data<T> operator()(F const& f, Data&& m, SubMatrix&& sm,
+                ir::node_data<std::int64_t>&& columns) const
+            {
+                std::unique_ptr<std::size_t[]> indices(
+                    std::make_unique<std::size_t[]>(columns.size()));
+
+                std::int64_t* p = columns.vector().data();
+                std::copy(p, p + columns.size(), indices.get());
+
+                auto result = blaze::columns(sm, indices.get(), columns.size());
+
+                return f.matrix(m, std::move(result));
+            }
+        };
+    }
+
     template <typename T, typename Data, typename F>
     ir::node_data<T> slice2d_basic_integer(Data&& m,
         ir::slicing_indices const& rows,
@@ -263,16 +308,11 @@ namespace phylanx { namespace execution_tree
         {
             HPX_ASSERT(row_stop > row_start);
 
-            auto sm = blaze::submatrix(m, row_start, 0ll,
-                row_stop - row_start, m.columns());
+            auto&& sm = blaze::submatrix(
+                m, row_start, 0ll, row_stop - row_start, m.columns());
 
-            static_assert(sizeof(std::int64_t) == sizeof(std::size_t),
-                "sizeof(std::int64_t) == sizeof(std::size_t)");
-            auto result = blaze::columns(sm,
-                reinterpret_cast<std::size_t*>(columns.vector().data()),
-                columns.size());
-
-            return f.matrix(m, std::move(result));
+            return detail::extract_columns<T, std::size_t>{}(
+                f, std::move(m), std::move(sm), std::move(columns));
         }
 
         // general case, pick arbitrary elements from matrix
@@ -379,6 +419,48 @@ namespace phylanx { namespace execution_tree
     }
 
     ///////////////////////////////////////////////////////////////////////////
+    namespace detail
+    {
+        template <typename T, typename Index, typename Enable = void>
+        struct extract_rows;
+
+        template <typename T, typename Index>
+        struct extract_rows<T, Index,
+            typename std::enable_if<sizeof(std::int64_t) == sizeof(Index)>::type>
+        {
+            template <typename F, typename Data, typename SubMatrix>
+            ir::node_data<T> operator()(F const& f, Data&& m, SubMatrix&& sm,
+                ir::node_data<std::int64_t>&& rows) const
+            {
+                auto result = blaze::rows(sm,
+                    reinterpret_cast<std::size_t*>(rows.vector().data()),
+                    rows.size());
+
+                return f.matrix(m, std::move(result));
+            }
+        };
+
+        template <typename T, typename Index>
+        struct extract_rows<T, Index,
+            typename std::enable_if<sizeof(std::int64_t) != sizeof(Index)>::type>
+        {
+            template <typename F, typename Data, typename SubMatrix>
+            ir::node_data<T> operator()(F const& f, Data&& m, SubMatrix&& sm,
+                ir::node_data<std::int64_t>&& rows) const
+            {
+                std::unique_ptr<std::size_t[]> indices(
+                    std::make_unique<std::size_t[]>(rows.size()));
+
+                std::int64_t* p = rows.vector().data();
+                std::copy(p, p + rows.size(), indices.get());
+
+                auto result = blaze::rows(sm, indices.get(), rows.size());
+
+                return f.matrix(m, std::move(result));
+            }
+        };
+    }
+
     template <typename T, typename Data, typename F>
     ir::node_data<T> slice2d_integer_basic(Data&& m,
         ir::node_data<std::int64_t> && rows,
@@ -450,13 +532,8 @@ namespace phylanx { namespace execution_tree
             auto sm = blaze::submatrix(m, 0ll, col_start,
                 m.rows(), col_stop - col_start);
 
-            static_assert(sizeof(std::int64_t) == sizeof(std::size_t),
-                "sizeof(std::int64_t) == sizeof(std::size_t)");
-            auto result = blaze::rows(sm,
-                reinterpret_cast<std::size_t*>(rows.vector().data()),
-                rows.size());
-
-            return f.matrix(m, std::move(result));
+            return detail::extract_rows<T, std::size_t>{}(
+                f, std::move(m), std::move(sm), std::move(rows));
         }
 
         // general case, pick arbitrary elements from matrix

